File: hot_hand.m
#import <Foundation/Foundation.h>

int flip() {
    return arc4random() % 2;
}

int fourFlips() {
    int flips[4] = { flip(), flip(), flip(), flip() };
    int result = 0;
    for (int i = 1; i < 4; i++) {
        if (flips[i-1] == 0 && flips[i] == 0) {
            result++;
        }
    }
    return result;
}

int millionFlips() {
    int result = 0;
    for (int i = 1; i <= 1000000; i++) {
        result += fourFlips();
    }
    return result;
}

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        printf("%d\n", millionFlips());
    }
    return 0;
}

File: hot_hand.js
#!/usr/bin/env node
// http://programmingpraxis.com/2015/10/02/the-hot-hand/

function fourFlips() {
  var flips = [flip(), flip(), flip(), flip()];
  var heads = 0; var result = 0;
  for (var i = 1; i < flips.length; i++) {
    if (flips[i-1] === heads && flips[i] === heads) {
      result += 1;
    }
  }
  return result;
}

function flip() {
  return Math.round(Math.random());
}

function oneMillionFlips() {
  var result = 0;
  for (var i = 1; i < 1000000; i++) {
    result += fourFlips();
  }
  return result;
}

console.log(oneMillionFlips());
